Browser_IDE/CodeEditor.js

```diff
@@ -0,0 +1,93 @@
+class CodeEditor {
+    constructor(filename,codearea,executionEnvironment, storedProject) {
+        this.filename = filename;
+        this.editorout =   this.setupCodeArea(codearea);
+        this.executionEnvironment = executionEnvironment;
+        this.storedProject = storedProject;
+
+    }
+
+    setupCodeArea(element){
+        let editor = CodeMirror.fromTextArea(element, {
+            mode: "text/javascript",
+            theme: "dracula",
+            lineNumbers: true,
+            autoCloseBrackets: true,
+            styleActiveLine: true,
+            extraKeys: {"Ctrl-Space": "autocomplete"},
+            hintOptions: {
+                alignWithWord: false,
+                completeSingle: false,
+                useGlobalScope: false,
+            },
+            foldGutter: true,
+            gutters: ["CodeMirror-linenumbers", "CodeMirror-foldgutter"]
+        });
+    
+        editor.on('inputRead', (cm, change) => {
+            if (!cm.state.completeActive) {
+                cm.showHint();
+            }
+        });
+        return editor;
+    }
+
+    // Utility function for saving/loading the code blocks
+    async  fileAsString(buffer){
+        return new Promise((resolve,error) => {
+            //_arrayBufferToString from https://stackoverflow.com/a/14078925
+            // Thanks Will Scott!
+            function _arrayBufferToString(buffer) {
+                var bb = new Blob([new Uint8Array(buffer)]);
+                var f = new FileReader();
+                f.onload = function(e) {
+                    resolve(e.target.result);
+                };
+                f.readAsText(bb);
+            }
+            if (typeof buffer === 'string' || buffer instanceof String)
+                resolve(buffer);
+            else
+                return _arrayBufferToString(buffer);
+        });
+    }
+
+    clearErrorLines(editors){
+        for (let editor of editors){
+            for (var i = 0; i < editor.lineCount(); i++) {
+                editor.removeLineClass(i, "wrap", "error-line");
+            }
+        }
+    }
+    
+
+    runCode(editors) {
+        this.clearErrorLines(editors);
+        let code = this.editorout.getValue();
+        this.executionEnvironment.runCodeBlock(this.filename, code);
+    }
+    
+
+    async saveCode(codePath,filePath) {
```

Comment:
i found when i move the codePath, and filePath (initCodePath and mainLoopCodePath) as a parameter into class, it will cause the error on console:"CodeEditor.js:77 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'filePath')
    at CodeEditor.js:77:42
    at async IDBStoredProject.access (IDBStoredProject.js:31:20)
    at async CodeEditor.saveCode (CodeEditor.js:75:9)" 
    but the function is fine,  I can print out the program, and run, pause , restart function is fine, and I modify some code inside the IDE, I can affect the result in the print out program. but there is still an error. and I cant find any code is defining the initCodePath, and mainLoopCodePath.
  so I think it Is better to left it there.